aws/ec2.go

```diff
@@ -0,0 +1,225 @@
+	return idDoc.InstanceID, nil
+}
+
+func (c *ec2MetadataClient) getEC2Region(ctx context.Context) (string, error) {
+	region, err := c.ec2MetadataSvc.RegionWithContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	log.Debugf("region: %s", region)
+	return region, nil
+}
+
+type ec2Client struct {
+	ec2Svc        *ec2.EC2
+	ec2Region     string
+	ec2InstanceId string
+}
+
+func NewEc2Client(awsSession *awssession.Session, region, instanceId string) (*ec2Client, error) {
+	ec2Svc := ec2.New(awsSession, aws.NewConfig().WithRegion(region))
+	if ec2Svc == nil {
+		return nil, fmt.Errorf("error connecting to EC2 service")
+	}
+
+	return &ec2Client{
+		ec2Svc:        ec2Svc,
+		ec2Region:     region,
+		ec2InstanceId: instanceId,
+	}, nil
+}
+
+func (c *ec2Client) getEC2NetworkInterfaceId(ctx context.Context) (networkInstanceId string, err error) {
+	input := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{
+			aws.String(c.ec2InstanceId),
+		},
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("network-interface.attachment.device-index"),
+				Values: []*string{
+					aws.String("0"), // Only device-id-0
+				},
+			},
+		},
+	}
+
+	i := 0
+retry:
+	out, err := c.ec2Svc.DescribeInstancesWithContext(ctx, input)
+	if err != nil {
+		if retriable(err) && i < defaultRetries {
+			log.WithField("instance-id", c.ec2InstanceId).Debug("retrying getting network-interface-id")
+			i++
+			goto retry
+		}
+		return "", err
+	}
+
+	if len(out.Reservations) <= 0 {
+		return "", fmt.Errorf("no network-interface-id found for EC2 instance %s", c.ec2InstanceId)
+	}
+
+	var interfaceId string
+	for _, instance := range out.Reservations[0].Instances {
+		if len(instance.NetworkInterfaces) == 0 {
+			return "", fmt.Errorf("no network-interface-id found for EC2 instance %s", c.ec2InstanceId)
+		}
+		for _, networkInterface := range instance.NetworkInterfaces {
+			if *(networkInterface.Attachment.DeviceIndex) == 0 {
+				interfaceId = *(networkInterface.NetworkInterfaceId)
+				if interfaceId != "" {
+					log.Debugf("instance-id: %s, network-interface-id: %s", c.ec2InstanceId, interfaceId)
+					return interfaceId, nil
+				}
+			}
+			log.Debugf("instance-id: %s, network-interface-id: %s", c.ec2InstanceId, interfaceId)
+		}
+		if interfaceId == "" {
+			return "", fmt.Errorf("no network-interface-id found for EC2 instance %s", c.ec2InstanceId)
+		}
+	}
+	return interfaceId, nil
+}
+
+func (c *ec2Client) setEC2SourceDestinationCheck(ctx context.Context, ec2NetId string, checkVal bool) error {
+	input := &ec2.ModifyNetworkInterfaceAttributeInput{
+		NetworkInterfaceId: aws.String(ec2NetId),
+		SourceDestCheck: &ec2.AttributeBooleanValue{
+			Value: aws.Bool(checkVal),
+		},
+	}
+
+	i := 0
+retry:
+	_, err := c.ec2Svc.ModifyNetworkInterfaceAttributeWithContext(ctx, input)
+	if err != nil {
+		if retriable(err) && i < defaultRetries {
+			log.WithField("net-instance-id", ec2NetId).Debug("retrying setting source-destination-check")
+			i++
+			goto retry
```

Comment:
Same here - let's ditch the `goto` and instead use a loop. That's not a pattern we typically use in Calico.